Implementation of Lichess download and extended filtering

Simske opened the pull request:
Closes #3 

Todo 
- [x] Lichess implementation
- [x] Implement Lichess-like filtering to chess.com

(Documentation moved into new PR #6)

chessdjw committed:
small fixes and restructering of project

Simske committed:
Implement basic lichess downloading

Simske committed:
Implemented date parsing

Simske committed:
Fixed subtract_month()

In case of current_month == months_to_subtract the year was not set
correctly, such that a date in the future was returned

Simske commented:
chess.com notes:
- [API documentation](https://www.chess.com/news/view/published-data-api)
- time control can be checked in JSON with `game["time_class"]`. `correspondence` for Lichess is `daily` for chess.com
- maybe also check `game["rules"] == "chess"` to ignore chess variants (for Lichess this is achieved with `perfType`)

Simske commented:
Do we want to include chess960 games?

chessdjw commented:
> Do we want to include chess960 games?

Rather not :D

chessdjw committed:
made Chess.com filters compatible with Lichess

Simske commented on pgn_downloader/chess_com.py:
```diff
@@ -1,36 +1,63 @@
+    username: str,
+    output_path,
+    color: str = None,
+    since: datetime = None,
+    until: datetime = None,
```
Why not
```python
since: datetime = datetime.min,
until: datetime = datetime.max,
```
This would eliminate the need for `is None` checks.

Simske commented on pgn_downloader/chess_com.py:
```diff
@@ -1,36 +1,63 @@
 
+    # Chess.com uses another terminology
+    for i in range(len(modes)):
+        modes[i] = modes[i].replace("correspondence", "daily")
```
Alternatively a list comprehension would make sense here. This also creates  copy, which could be useful if we ever want to run on both lichess and chess.com.
```python
modes = ["daily" if mode == "correspondence" else mode for mode in modes]
```

Simske commented on pgn_downloader/chess_com.py:
```diff
@@ -1,36 +1,63 @@
+                    continue
+            if until is not None:
+                if too_young(archive_date, until):
+                    continue
```
For the `is None` checks see comment in function definition.
Couldn't you replace `too_old` and `too_young` with a direct check?
```python
if end_of_month(archive_date) < since or archive_date > until:
    continue
```
with (`end_of_month()` from `pgn_downloader.date_parser`)

Simske commented on pgn_downloader/chess_com.py:
```diff
@@ -1,36 +1,63 @@
-    for month_url in archives:
-        print(f"Downloading games from {month_url[-7:]}", end="\r")
-        games = requests.get(month_url).json()["games"]
+    with open(output_path, "w") as f:
```
File should be opened in mode `x` for exclusive creation, to prevent existing files to be overwritten

chessdjw committed:
small code improvements in chess_com.py

chessdjw commented:
I followed your suggestions. One more question? Should I remove the game counting and console output about which archive is being downloaded? In Lichess we don't have it, so in case we want to make it more similar... On the other hand, the program runs considerably longer for Chess.com and it's nice to see that something is happening.

Simske commented:
I would leave it in. Maybe I'll add something similar (at least the game counting) to the Lichess part later.

Simske committed:
Customized CLI output, prevent full Exception stacktrace to print

A full exception traceback can intimidate user, only the exception
message should be enough.

Simske committed:
Fixed date comparisons in chess_com.py

chessdjw reviewed (approved):
Looks good! 👍
